### ISSUE TITLE
Trust factor issues in CSGO

### ISSUE THREAD

**Issue by KeironO** [flathub/com.valvesoftware.Steam#689]
**Game information**
Counter Strike: Global Offensive

**Distribution name and version where applicable**
```
⋊> ~ cat /etc/redhat-release                                           08:32:52
Fedora release 33 (Thirty Three)
```

**Flatpak info**
```
⋊> ~ flatpak --version                                                 08:32:53
Flatpak 1.8.4
⋊> ~ flatpak --gl-drivers                                              08:33:14
default
host
```

**Problem description**
I'm of the belief that the flatpak version of Steam is causing issues with trust factor in CSGO. I have a 10+ year old account with > 100 games, and have a commendation rate of 5% per game. I've checked my friends list for VAC-banned accounts, and there aren't any - and I have never knowingly queued with cheaters.

I was told yesterday, when starting a competitive MM search that my trust factor I was informed by my friend that they had a message about "Your matchmaking experience may be slightly impacted because Trust Factor of n0r13k is lower than yours.". I found this surprising, so after the game (which contained obvious cheaters) I ran flatpak update, which updated my flatpak instance - and when I queued again he was informed that "Your matchmaking experience will be significantly affected because Trust Factor of n0r13k is substantially lower than yours."

My Steam account can be found on: https://steamcommunity.com/id/kto456dog/

**Does this issue reproduce with native Steam**
I'm struggling to work out a way I can do this.

**Comment by guihkx** [comment]
Well, judging by the fact that the newly-introduced [Trusted Mode](https://support.steampowered.com/kb_article.php?ref=8835-KLXV-0292#trust) of CS:GO is strictly a Windows-only thing, it makes complete sense that Linux users will have their trust score negatively impacted.

**Comment by KeironO** [comment]
So Trusted Mode isn't enabled on Linux?

When I run `trusted_launch_info` in the console it does say it's enabled.

**Comment by guihkx** [comment]
When it was first announced, Trusted Mode [wasn't even enabled by default on Linux](https://github.com/ValveSoftware/csgo-osx-linux/issues/2482). While you could still launch the game with `-trusted` and get the "Trusted Mode experience", it was only [after a month](https://github.com/ValveSoftware/csgo-osx-linux/issues/2482#issuecomment-670279638) that Valve _silently_ "enabled" Trusted Mode on Linux by default.

In my opinion (because I haven't reversed CS:GO's Linux binaries to actually find out), Trusted Mode on Linux is just snake oil. Because a "Trusted Mode" would be effectively useless on Linux to being with, since you can go as far as modify the kernel.

Currently, the way Trusted Mode works on Windows is by hooking into `NtOpenFile()` and then blocking any file that is not part of CS:GO from injecting in its process. This was easily bypassed by cheat developers since day one, by the way.

Funnily enough, though, it's the legitimate players that [can't stream/record CS:GO using OBS](https://github.com/obsproject/obs-studio/issues/3139) anymore. 🤣 Not unless you run CS:GO with `-allow_third_party_software`, which may or may not hurt your trust score.

**Comment by KeironO** [comment]
> In my opinion (because I haven't reversed CS:GO's Linux binaries to actually find out), Trusted Mode on Linux is just snake oil. Because a "Trusted Mode" would be effectively useless on Linux to being with, since you can go as far as modify the kernel.

So do you think that Linux users are just being automatically carried into the "low trust factor" category because `NtOpenFile()` simply wouldn't work on Linux?

**Comment by guihkx** [comment]
Perhaps not automatically, but I do think that if you get reported by someone while playing on Linux, that report will "weight more" on you than if you were playing with trusted mode on Windows, therefore having a higher impact on your trust factor. But again, that's just speculation on my part and I can't confirm anything.

But it _does make sense_ when you consider that once upon a time, an **open-source**, Linux-only, CS:GO cheat (which I won't name) went undetected for a whole year before the Valve Anti-Cheat took action against people using it.

**Comment by KeironO** [comment]
Without making light of it, my individual performance in games are unlikely to reflect in someone suspecting that I'm cheating.

**Comment by guihkx** [comment]
Hahah you never know, man. I've been playing this game for 6 years now, and one thing I noticed now that I'm a better player, is that people in lower ranks will actually report you more than the ones in higher ranks (not talking about blatant cheaters, of course).

Regardless, [as Valve said themselves](https://blog.counter-strike.net/index.php/the-trust-factor/), reports are not the only thing that weighs into your trust score, so it could be something else outside of CS:GO that's influencing your score, who knows.

So, yeah... I enjoy playing Valve's matchmaking too, but when I do, I make sure to reboot into Windows, because personally I feel the Linux version is neglected by Valve and overall experience is better on Windows, which is sad.

**Comment by Newbytee** [comment]
> Well, judging by the fact that the newly-introduced [Trusted Mode](https://support.steampowered.com/kb_article.php?ref=8835-KLXV-0292#trust) of CS:GO is strictly a Windows-only thing, it makes complete sense that Linux users will have their trust score negatively impacted.

Baseless claim.

> My advice: Don't play CS:GO on Linux. At all. It's heavily neglected by Valve and I don't doubt that Valve will drop support for it in the near future.

Why do you say so?

**Comment by guihkx** [comment]
>Baseless claim.

Well, the closest thing to "DLL Injection" we have on Linux would be using `LD_PRELOAD`. And CS:GO can't do anything to prevent these libraries from loading. I'd imagine more sophisticated cheats would use `dlopen()` / `ptrace()` instead, though.

>Why do you say so?

Valve might not "officially" drop support for Linux, but in practice that's basically what happens today. A bunch of Linux-only issues being ignored for _years_. Some of these are so easily fixable that it came to a point that a player had to implement a downloader for community maps, after Linux users were ignored for **4 years**:

https://github.com/ericek111/linux-csgo-downloadfixer

Nowadays, though, you still have issues like these being ignored:

https://github.com/ValveSoftware/csgo-osx-linux/issues/1757 (game-breaking)
https://github.com/ValveSoftware/csgo-osx-linux/issues/2094
https://github.com/ValveSoftware/csgo-osx-linux/issues/2558 (game-breaking)
https://github.com/ValveSoftware/csgo-osx-linux/issues/1958
https://github.com/ValveSoftware/csgo-osx-linux/issues/1929

None of these are present on Windows, so that's why I have to play there.

**Comment by KeironO** [comment]
I've posted this issue on reddit, and it sounds like a flatpak specific issue.

https://old.reddit.com/r/linux_gaming/comments/krkzlf/is_csgos_trusted_launch_broken_on_linux/

**Comment by nanonyme** [comment]
Could be some seccomp incompatibility.

**Comment by tilosp** [comment]
this is definitely not a general flatpak problem since I am using the flatpak and my trust factor is fine

**Comment by henrym11106** [comment]
> Nowadays, though, you still have issues like these being ignored:
> 
> [ValveSoftware/csgo-osx-linux#1757](https://github.com/ValveSoftware/csgo-osx-linux/issues/1757) (game-breaking)
> [ValveSoftware/csgo-osx-linux#2094](https://github.com/ValveSoftware/csgo-osx-linux/issues/2094)
> [ValveSoftware/csgo-osx-linux#2558](https://github.com/ValveSoftware/csgo-osx-linux/issues/2558) (game-breaking)
> [ValveSoftware/csgo-osx-linux#1958](https://github.com/ValveSoftware/csgo-osx-linux/issues/1958)
> [ValveSoftware/csgo-osx-linux#1929](https://github.com/ValveSoftware/csgo-osx-linux/issues/1929)
> 
> None of these are present on Windows, so that's why I have to play there.

1757 is only game breaking if you want to use a gamepad to play on linux. it also seems to cause stuttering on windows? just use -nojoy

the other "game breaking" issue 2558 was closed as a duplicate of [2483 where you wrote](https://github.com/ValveSoftware/csgo-osx-linux/issues/2483#issuecomment-710428075) 
> This doesn't seem to be Linux-exclusive, as I'm experiencing it on both Windows 10 and Arch Linux
this issue I want to explore because on my main computer I've never had it but another machine in my home does from time to time despite running the same software and AMD GPU while connected to the same server.

anyway my experience with Arch Linux tends to be more pleasant than Windows10 and nowadays the performance tends to be significantly better, too. my main issues right now are:
https://github.com/ValveSoftware/csgo-osx-linux/issues/2102 (AMD GPUs only?)
and the fact that FACEIT doesn't want to make an anticheat client for Linux as far as I can tell.

given the performance and licensing benefits of Linux I can't see Valve dropping CSGO Linux, they do run all their CSGO servers on Linux as do some enthusiast server operators like TarikGG

**Comment by henrym11106** [comment]
> Nowadays, though, you still have issues like these being ignored:
> 
> [ValveSoftware/csgo-osx-linux#1757](https://github.com/ValveSoftware/csgo-osx-linux/issues/1757) (game-breaking)
> [ValveSoftware/csgo-osx-linux#2094](https://github.com/ValveSoftware/csgo-osx-linux/issues/2094)
> [ValveSoftware/csgo-osx-linux#2558](https://github.com/ValveSoftware/csgo-osx-linux/issues/2558) (game-breaking)
> [ValveSoftware/csgo-osx-linux#1958](https://github.com/ValveSoftware/csgo-osx-linux/issues/1958)
> [ValveSoftware/csgo-osx-linux#1929](https://github.com/ValveSoftware/csgo-osx-linux/issues/1929)
> 
> None of these are present on Windows, so that's why I have to play there.

1757 is only game breaking if you want to use a gamepad to play on linux. it also seems to cause stuttering on windows? just use -nojoy

the other "game breaking" issue 2558 was closed as a duplicate of [2483 where you wrote](https://github.com/ValveSoftware/csgo-osx-linux/issues/2483#issuecomment-710428075) 
> This doesn't seem to be Linux-exclusive, as I'm experiencing it on both Windows 10 and Arch Linux

this issue I want to explore because on my main computer I've never had it but another machine in my home does from time to time despite running the same software and AMD GPU while connected to the same server.

anyway my experience with Arch Linux tends to be more pleasant than Windows10 and nowadays the performance tends to be significantly better, too. my main issues right now are:
https://github.com/ValveSoftware/csgo-osx-linux/issues/2102 (AMD GPUs only?)
and the fact that FACEIT doesn't want to make an anticheat client for Linux as far as I can tell.

given the performance and licensing benefits of Linux I can't see Valve dropping CSGO Linux, they do run all their CSGO servers on Linux as do some enthusiast server operators like TarikGG

**Comment by guihkx** [comment]
>1757 is only game breaking if you want to use a gamepad to play on linux.

No. CS:GO will crash if you have a gamepad paired/plugged-in. You either have to use `-nojoy` or disconnect the gamepad, which is stupid.

> the other "game breaking" issue 2558 was closed as a duplicate of [2483 where you wrote](https://github.com/ValveSoftware/csgo-osx-linux/issues/2483#issuecomment-710428075)

That issue was closed as duplicate incorrectly, IMO. Because the SteamNetworkingSockets error happens regardless of game mode, i.e. competitive, casual, and the worse of them all: danger zone. It might not matter that much in lower ranks, but having played a competitive match on Linux where that error pops up literally _every second_, versus on Windows where it doesn't happen at all (with both OSes in the same machine, by the way), I _really_ do think it's a game-breaking bug.

>anyway my experience with Arch Linux tends to be more pleasant than Windows10 and nowadays the performance tends to be significantly better, too. my main issues right now are:

I use Arch and Windows 10 as well, but unfortunately I have the opposite view. For starters, mouse input just feels much smoother on Windows 10 than on GNOME (which is what I use), although this might be GNOME-specific because when I used KDE it felt just like Windows; Also no annoying SteamNetworkingSockets errors; Downloading demos won't crap on fps; No stupid xrandr commands to play properly on 4:3 stretched, but even then, you end up with an [annoying color-fringed, ghost crosshair](https://i.imgur.com/Kz1Ych1.jpg).

The list goes on...

>they do run all their CSGO servers on Linux

I'm pretty sure they mainly do it because it wouldn't make financial sense to pay for thousands of Windows Server licenses just so they host their servers, lol.

**Comment by henrym11106** [comment]
This is getting a bit off-topic but I am in awe. I never imagined that color fringe issue could even be possible. 4:3 stretch is default CSGO behavior on every Linux system I've used regardless of DE and hardware. something very weird is afoot.

Mouse input is another improvement on Linux for me because on windows it is effectively increased and affected by aspect ratio while in-game at 1024x768 stretch. it's completely smooth and consistent on every machine I've used with Arch and Debian-based OS + Gnome, though I normally use KDE.

I'm beginning to think something is up with either your Arch install or hardware.
